src/routes/download/download.service.ts

```diff
@@ -0,0 +1,125 @@
+import { get } from 'lodash';
+
+import getAppConfig from '../../config/global';
+import logger from '../../logger';
+import { getEsClient } from '@/esClient';
+import { downloadFromAccessUrl, DRSMetada, fetchDRSMetadata, parseDRSPath } from './drs.service';
+import baseConfig from '../../../configs/base.json';
+
+const retrieveFiles = async (drsInfo: { sequenceId: string; drsPath: string }[]) => {
+  // download the https access urls for each file
+  const sequenceInfo = await Promise.all(
+    drsInfo.map(async (drsUrl) => {
+      const accessUrlInfo = await fetchDRSMetadata(drsUrl.drsPath);
+      return {
+        sequenceId: drsUrl.sequenceId,
+        accessUrlInfo,
+      };
+    }),
+  );
+
+  const downloaded = sequenceInfo.map(
+    async (sequence: { sequenceId: string; accessUrlInfo: DRSMetada }) => {
+      try {
+        // retrieve file from access url
+        // return from first url that resolves
+        const stream = (await Promise.any(
+          sequence.accessUrlInfo.accessUrls.map((u: any) =>
+            downloadFromAccessUrl(u.access_url.url),
+          ),
+        )) as unknown as NodeJS.ReadableStream;
+        logger.info(`returning stream for ${sequence.accessUrlInfo.name}`);
+        return {
+          name: sequence.accessUrlInfo.name,
+          stream,
+        };
+      } catch (err) {
+        if (err instanceof AggregateError) {
+          logger.error(
+            `Multiple errors reported downloading from access url for ${sequence.sequenceId}: ${err.errors}`,
+          );
+          throw err.errors[0];
+        } else {
+          logger.error(`Error on ${sequence.sequenceId}: ${err}`);
+          throw err;
+        }
+      }
+    },
+  );
+  const files = await Promise.all(downloaded);
+  return files;
+};
+
+const fetchSequenceData = async (ids: string[]) => {
+  const esClient = await getEsClient();
+  const response: any = await esClient
+    .search({
+      index: baseConfig.index,
+      body: {
+        query: {
+          bool: {
+            filter: {
+              terms: {
+                sequence_id: ids,
+              },
+            },
+          },
+        },
+      },
+    })
+    .catch((err) => {
+      logger.error('ES query failed: ', err);
+      throw new Error('ES query failed, cannot retrieve sequence file info.');
+    });
+
+  const queryResults: any[] = get(response, 'body.hits.hits', []).map((result: any) => {
+    const files = get(result, '_source.files');
+    return {
+      sequenceId: get(result, '_id'),
+      files,
+    };
+  });
+  return queryResults;
+};
+
+const appendEmptySequence = (sequenceId: string, fileResults: any[]) =>
+  fileResults.concat({ sequenceId, files: [] });
+
+export const downloadSequenceFiles = async (ids: string[]) => {
+  const config = await getAppConfig();
+  if (!ids.length || ids.length > config.download.sequences_limit) {
+    throw new Error('Invalid number of sequences requested for download!');
+  }
```

Comment:
This seems like the responsibility of the router to validate the inputs provided. It should be able to return more relevant text to the UI this way:
0 = no sequence IDs provided
>max = too many sequences requested, maximum possible is ${max}

Reply:
you can leave this error here to ensure nothing ever calls this method with too many ids, but input validation with good messaging for the UI should be added to the router.